Fix blank code coverage 1621

martinisaksenprogleasing opened the pull request:
## PR Summary
Adding ability to recursively traverse sub folders to find all files to be included in code coverage. This can be turned off by setting a new configuration option (CodeCoverageConfiguration.RecursePaths) to false. I have set it to true by default as I feel that this is the default expected behavior.

I also noticed that the coverage.xml file was not fully compatible with ReportGenerator and so I updated the sourcefile names to include the relative path so the code could be populated correctly in the finished report.

This new option should be documented at the https://pester.dev/docs/usage/code-coverage

`Fix #1621` 

## PR Checklist

- [x] PR has meaningful title
- [x] Summary describes changes
- [x] PR is ready to be merged
  - If not, use the arrow next to `Create Pull Request` to mark it as a draft. PR can be marked `Ready for review` when it's ready.
- [x] Tests are added/update *(if required)*
- [ ] Documentation is updated/added *(if required)*

nohwnd commented on src/functions/Coverage.ps1:
```diff
@@ -107,30 +107,24 @@ function Convert-UnknownValueToInt {
-        $resolvedPaths = & $SafeCommands['Resolve-Path'] -Path $path -ErrorAction Stop |
-            & $SafeCommands['Where-Object'] { $includeTests -or $_.Path -notlike $testsPattern }
+        foreach ($path in $paths) {
+            $resolvedPaths += & $SafeCommands['Resolve-Path'] -Path $path -ErrorAction Stop |
```
This is slow, you should output from foreach directly into the variable, or use List and Add. There is a factory method on Pester Factory type to create a list.

nohwnd commented on src/functions/Coverage.ps1:
```diff
@@ -145,6 +139,33 @@ function Resolve-CoverageInfo {
+        if ($item -is [System.IO.FileInfo] -and ('.ps1', '.psm1') -contains $item.Extension) {
+            $item.FullName
+        }
+        elseif ($item -is [System.IO.DirectoryInfo] -and $PesterPreference.CodeCoverage.RecursePaths) {
```
This must use .Value

nohwnd reviewed (changes requested):
Please add tests, I don't think the change works as is.

nohwnd commented:
Before you implement the changes mentioned in the review, can we talk about why this is needed? Do you have a specific case where this change is beneficial? Does this change the current behavior?

martinisaksenprogleasing commented:
This is to fix #1621 and, as mentioned in the summary, it will change the default behavior for `-CI` to traverse through subfolders when looking for files for code coverage.
I was having the same issue of an empty code coverage file and running the proposed code changes gave me the correct code coverage information.

martinisaksenprogleasing commented:
@nohwnd I updated the code and tests but the pipeline failed in PS4 for `Demo -Run only what is needed`. 

https://nohwnd.visualstudio.com/Pester/_build/results?buildId=1304&view=logs&j=ff75994e-ad5b-5649-588c-530e2c306b35&t=bb64baaf-a3d6-5598-2b44-c9bedcc4901a&l=1815

Not sure how my changes would affect this, unless it is spending too much time finding code coverage files. But that tests doesn't have code coverage enabled...

Any ideas?

martinisaksenprogleasing commented:
@nohwnd ☝️

martinisaksenprogleasing reviewed (commented):
Updated code based on feedback

nohwnd commented:
> Not sure how my changes would affect this, unless it is spending too much time finding code coverage files. But that tests doesn't have code coverage enabled...

It is not your fault IMHO.

nohwnd commented:
You commented that you addressed the feedback but there is no new commit. Did you push all your changes?

martinisaksenprogleasing commented:
> You commented that you addressed the feedback but there is no new commit. Did you push all your changes?

Yes. I see them in the commit history and in the files changed diff. Are they not showing up for you?

martinisaksenprogleasing commented:
@nohwnd The changes have been committed. Is there anything else that needs to be done?

martinisaksenprogleasing commented:
@nohwnd All the changes are there. Can you take another look please?
